src/plugins/discover/public/application/main/hooks/use_adhoc_data_views.ts

```diff
@@ -57,22 +69,27 @@ export const useAdHocDataViews = ({
         prev.filter((d) => d.id && dataViewToUpdate.id && d.id !== dataViewToUpdate.id)
       );
 
-      savedSearch.searchSource.setField('index', newDataView);
-
-      // update filters references
       const uiActions = await getUiActions();
       const trigger = uiActions.getTrigger(UPDATE_FILTER_REFERENCES_TRIGGER);
       const action = uiActions.getAction(UPDATE_FILTER_REFERENCES_ACTION);
+
+      // execute shouldn't be awaited, this is important for pending history push cancellation
       action?.execute({
         trigger,
         fromDataView: dataViewToUpdate.id,
         toDataView: newDataView.id,
         usedDataViews: [],
       } as ActionExecutionContext);
+      stateContainer.kbnUrlStateStorage.kbnUrlControls.flush(true);
+
+      const updatedFilters = filterManager.getFilters();
+      savedSearch.searchSource.setField('index', newDataView);
+      stateContainer.replaceUrlAppState({ index: newDataView.id, filters: updatedFilters });
```

Comment:
Thinking about this, since it sounds like a potential race condition. 
Since we have the filters in our AppState, couldn't we get those, update the indices , use replaceUrlAppState to update? Like this, just pseudo code here:

```suggestion
      const currentState = stateContainer.appStateContainer.getState();

      const updatedFilters = updateFilterDataViewId(currentState.filters);
      savedSearch.searchSource.setField('index', newDataView);
      stateContainer.replaceUrlAppState({ index: newDataView.id, filters: updatedFilters });
```

update the indices there, so we wouldn't need to do the URL flushing?

Reply:
I've double checked this and found that URL flushing is not needed here. So, replacing id would be enough.

Reply:
We anyway need to trigger `setFilters` according to a normal flow. It's done by the update data view id action.

Reply:
Fine with that, thx 👍